File: 131.palindrome-partitioning.cpp
/*
 * @lc app=leetcode id=131 lang=cpp
 *
 * [131] Palindrome Partitioning
 */
#include <string>
#include <vector>

using namespace std;

// @lc code=start
class Solution {
public:
    // T: O(n * 2 ^ n)
    // S: O(n ^ 2) for the is_palindrome 2D array
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        const int n = s.size();
        
        // is_palindrome[i][j] means whether s[i:j+1] is a palindrome.
        // Precompute this dp info to reduce time cost in backtracking.
        // In Q132, this part is essential as backtracking is not needed.
        vector<vector<uint8_t>> is_palindrome(n, vector<uint8_t>(n, 0));  // avoid using vector<bool>
        for (int r = 0; r < n; ++r) {
            for (int l = 0; l <= r; ++l) {
                if (s[l] == s[r] && (r - l <= 2 || is_palindrome[l + 1][r - 1])) {
                    is_palindrome[l][r] = 1;
                }
            }
        }

        vector<string> parts;
        backtrack(s, parts, 0, is_palindrome, res);
        return res;
    }

private:
    void backtrack(const string &s, vector<string> &parts, int begin, const vector<vector<uint8_t>> &is_palindrome, vector<vector<string>> &res) {
        if (begin == s.size()) {   
            // Append a copy
            res.push_back(parts);
            return;
        }

        for (int end = begin; end < s.size(); ++end) {
            if (is_palindrome[begin][end]) {
                parts.push_back(s.substr(begin, end - begin + 1));
                backtrack(s, parts, end + 1, is_palindrome, res);
                parts.pop_back();
            }
        }
    }
};
// @lc code=end

File: 131.palindrome-partitioning.go
/*
 * @lc app=leetcode id=131 lang=golang
 *
 * [131] Palindrome Partitioning
 */
package lc

// @lc code=start
func backtrack(s string, begin int, isPalindrome [][]bool, parts *[]string, res *[][]string) {
	if begin == len(s) {
		*res = append(*res, append([]string{}, *parts...))
		return
	}

	for i := begin; i < len(s); i++ {
		if isPalindrome[begin][i] {
			*parts = append(*parts, s[begin:i+1])
			backtrack(s, i+1, isPalindrome, parts, res)
			*parts = (*parts)[:len(*parts)-1]
		}
	}
}

// T: O(n * 2 ^ n)
// S: O(n ^ 2)
func partition(s string) [][]string {
	n := len(s)

	// dp[i][j] means whether s[i:j+1] is a palindrome.
	dp := make([][]bool, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]bool, n)
	}
	for r := 0; r < n; r++ {
		for l := 0; l <= r; l++ {
			if s[l] == s[r] && (r-l <= 2 || dp[l+1][r-1]) {
				dp[l][r] = true
			}
		}
	}

	var res [][]string
	var parts []string
	backtrack(s, 0, dp, &parts, &res)
	return res
}

// @lc code=end
